src/PrestaShopBundle/Controller/Admin/Improve/International/TaxRulesGroupController.php

```diff
@@ -0,0 +1,657 @@
+        } catch (Exception $e) {
+            return $this->redirectToRoute('admin_tax_rules_groups_index');
+        }
+
+        try {
+            $taxRuleFormBuilder = $this->get(
+                        'prestashop.core.form.identifiable_object.builder.tax_rule_form_builder'
+                    );
+
+            $taxRuleFormHandler = $this->get(
+                        'prestashop.core.form.identifiable_object.handler.tax_rule_form_handler'
+                    );
+
+            $taxRuleForm = $taxRuleFormBuilder->getFormFor(
+                        $taxRuleId,
+                        [
+                            'taxRulesGroupId' => $editableTaxRule->getTaxRulesGroupId()->getValue(),
+                            'taxRuleId' => $taxRuleId,
+                        ]
+                    );
+
+            $taxRuleForm->handleRequest($request);
+
+            $result = $taxRuleFormHandler->handleFor($taxRuleId, $taxRuleForm);
+
+            if ($result->isSubmitted() && $result->isValid()) {
+                $this->addFlash('success', $this->trans('Successful update.', 'Admin.Notifications.Success'));
+            }
+
+            if ($result->isSubmitted() && !$result->isValid()) {
+                $this->setTaxRuleSessionParameters(
+                            $this->generateUrl('admin_tax_rules_edit', ['taxRuleId' => $taxRuleId]),
+                            $taxRuleId
+                        );
+
+                return $this->redirectToRoute('admin_tax_rules_groups_edit', [
+                            'taxRulesGroupId' => $editableTaxRule->getTaxRulesGroupId()->getValue(),
+                            'request' => $request,
+                        ], 307);
+            }
+        } catch (Exception $e) {
+            $this->addFlash('error', $this->getErrorMessageForException($e, $this->getErrorMessages($e)));
+        }
+
+        return $this->redirectToRoute('admin_tax_rules_groups_edit', [
+            'taxRulesGroupId' => $editableTaxRule->getTaxRulesGroupId()->getValue(),
+        ]);
+    }
+
+    /**
+     * Loads tax rule data for ajax requests
+     *
+     * @AdminSecurity(
+     *     "is_granted('update', request.get('_legacy_controller'))",
+     *     redirectRoute="admin_tax_rules_groups_index",
+     * )
+     *
+     * @param int $taxRuleId
+     *
+     * @return Response
+     */
+    public function loadTaxRuleAction(int $taxRuleId)
+    {
+        try {
+            /** @var EditableTaxRule $editableTaxRule */
+            $editableTaxRule = $this->getQueryBus()->handle(new GetTaxRuleForEditing((int) $taxRuleId));
+            $statesProvider = $this->get('prestashop.adapter.form.choice_provider.country_state_by_id');
+            $states = $statesProvider->getChoices([
+                'id_country' => $editableTaxRule->getCountryId()->getValue(),
+            ]);
+            $normalizer = new ObjectNormalizer(null, new CamelCaseToSnakeCaseNameConverter());
```

Comment:
I like that 👍

Reply:
Would it make sense to wrapp these 2 services ($normalizer and $serializer) into `ResponseSerializer` or something? So we don't need to initialize them on our own.

Reply:
If we use it more, yes 👍

Reply:
I added definition of json serializer: https://github.com/PrestaShop/PrestaShop/blob/d4f6aa05caea86f1d3363accc14829eb330760b9/src/PrestaShopBundle/Resources/config/services/bundle/serializer.yml#L13. But i guess we should wait for https://github.com/PrestaShop/PrestaShop/issues/15977#issue-507785941 to be resolved